Fix Cli hanging on error, don't manually kill process

kuhnroyal opened the pull request:
Not sure why this is required, the process should complete on it's own.
There are no leftover processes with or without errors/lints.

kuhnroyal commented:
Having no problems locally with this change, as far as I can tell.

rrousselGit commented:
I don't like this change. It's effectively removing some cleanup logic
We should instead figure out why that line hangs rather than removing it.

kuhnroyal commented:
I agree, but I spent 2 hours trying to figure it out. This issue is a big problem for using `custom_lint` in CI.

rrousselGit commented:
Why so? It should only impact custom_lint's CI right? User code shouldn't be impacted.

kuhnroyal commented:
As soon as you have some sort of error, which can happen, this is stuck in CI until it times out and the process is never killed even after the workflow timeout. This is not a big problem on Github hosted runners because they are killed off after the timeout. But private hosted runners or other CI systems where the instance stays alive have a problem with this.

rrousselGit commented:
Right but then you found a bug in custom_lint. That error shouldn't happen.
So I don't think it's really a big deal since I'm the primary person impacted.


It'd obviously be better to fix it. But I don't think that's the fix.

kuhnroyal commented:
This also happens when there is an error the users own plugin, and that is a very likely scenario.
Some null access in your own lint and the process is stuck atm.

rrousselGit commented:
Now _that's_ a good reason :)
Didn't know about that one. Fair enough then

rrousselGit commented:
Do you have a specific scenario where an error causes such behavior?
I tried throwing inside a lint but I don't have a problem.

kuhnroyal commented:
I will check this again, I specifically had this problem. We use private Github runners and I had a lot of leftover processes.

kuhnroyal commented:
The `Completer<Process>` never completed because the `init()` Future crashed before it was linked to the completer.

rrousselGit commented on packages/custom_lint/lib/src/v2/server_to_client_channel.dart:
```diff
@@ -409,27 +409,29 @@ $dependencies
-          await _serverSocket.then((value) => value.port.toString())
-        ],
-        workingDirectory: _tempDirectory.path,
+    final processFuture = Future.sync(() {
```
At that point, the logic should be extracted to a separate async function. That'd avoid this funky `Future`

kuhnroyal replied:
Done

rrousselGit commented:
Do you mind adding a test for this?

kuhnroyal commented:
Ok when writing the test I noticed that this was a bit more involved than I thought.
I reverted the funky future, the errors are correctly propagate already.
Instead we can just check if the completer has completed in `close()`.

There is no need to complete the Completer with the errors from the `Process`, since it is awaited they are also thrown to the `init()` caller, completing the `Completer<Process>` with the error would cause `close()` to rethrow the error again, this would cause further problems in the plugin that calls `close`.

kuhnroyal commented on packages/custom_lint/lib/src/v2/server_to_client_channel.dart:
```diff
@@ -222,15 +222,7 @@ void _writePackageConfigForTempProject(
-        'No package_config.json found. Did you forget to run `pub get`?\n'
-        'Tried to look in:\n${contextRoots.map((e) => '- ${e.root}\n').join()}',
-      );
-    }
```
This check is already done in `getPackageListForContextRoots`, it can never fail here.

kuhnroyal commented on packages/custom_lint/lib/src/v2/server_to_client_channel.dart:
```diff
@@ -409,6 +400,7 @@ $dependencies
 
   @override
   Future<void> init() async {
+    _packages = getPackageListForContextRoots(_contextRoots.roots);
```
I moved this here as this can throw in the same way as the other initialization calls. The packages are only used during the creation of the temp project.

rrousselGit commented on packages/custom_lint/lib/src/v2/server_to_client_channel.dart:
```diff
@@ -503,7 +488,7 @@ $dependencies
       _tempDirectory.delete(recursive: true),
       _serverSocket.then((value) => value.close()),
-      _process.future.then((value) => value.kill()),
+      if (_process.isCompleted) _process.future.then((value) => value.kill()),
```
Why this?

kuhnroyal replied:
Well the `Completer` should have been completed in `init`. If it wasn't because of errors thrown before it was completed, then this will wait forever (the hanging).

> There is no need to complete the Completer with the errors from the Process, since it is awaited they are also thrown to the init() caller, completing the Completer<Process> with the error would cause close() to rethrow the error again, this would cause further problems in the plugin that calls close.

And further because of this, we don't need the errors in the `Completer`, this is what the funky future did before.

rrousselGit replied:
I've added an onError instead. It probably makes more sense.  
And removed the complete for a `late final Future<T>`. So if somehow we forget to assign the future, `close` would throw instead of hang.

kuhnroyal replied:
Yea I guess, was thinking the same but thought you must have had your reasons for the completer.

rrousselGit commented on packages/custom_lint/test/cli_test.dart:
```diff
@@ -327,4 +328,43 @@ void main() {
+
+    await runWithIOOverride(
+      (out, err) async {
+        await cli.entrypoint();
```
Could you use `Process.run` instead of `cli.entrypoint` to achieve this?
This would test that the CLI correctly exits.

Technically this test doesn't guarantee that the command indeed stops; only that it sets the exitCode and logs

rrousselGit replied:
You wouldn't need runWithIOOverride either then

rrousselGit replied:
We'd also want a similar test, but for when there's a dependency conflict error.

kuhnroyal replied:
Yea I just tried this but this immediately fails with dependency conflicts on `vm_service` which is constrained to `<10.0.0` in `custom_lint_builder` via `hotreloader`. Not exactly sure how to solve this, probably need to add constraints to the generated pusbspec.yaml

rrousselGit replied:
What does it conflict with?

kuhnroyal replied:
> We'd also want a similar test, but for when there's a dependency conflict error.

I added a new integration test for this as well but I don't really like the way the conflict is produced. I don't see another way tho.

kuhnroyal commented:
This now rebased on https://github.com/invertase/dart_custom_lint/pull/121
I added a separate integration test for 4 different cases of process termination.
When running `dart run custom_lint` via a process I can't find a way to prevent it from running `pub get`. Additionally there are dependency overrides required because the `pubspec_overrides.yaml` are not taken into account, this generates warnings in the test which I skipped in the test output.

kuhnroyal commented on packages/custom_lint/test/create_project.dart:
```diff
@@ -279,14 +293,21 @@ Directory createDartProject({
+  if (parent == null) {
+    newFolder = Directory.systemTemp.createTempSync(name);
+  } else {
+    newFolder = Directory(join(parent.path, name))..createSync();
```
I added the option to pass a parent directory all the way down here, otherwise there is no easy way to create multiple context roots, wich are required for the two tests that lead to the hanging process.

kuhnroyal commented:
I have no idea why there are no dependency override warnings in the output here on the Github runners.

rrousselGit commented on packages/custom_lint/test/cli_process_test.dart:
Thanks for this!

rrousselGit commented:
It's looking pretty good! Thanks for all your work :)
